### PR TITLE
Refactor Stock distribution

### LINKED ISSUE #1854: (refactor) Stock distribution
# TO PATIENT

1. Strange behavior when the user provide a bad patient code or name (mostly related to the component it self)

2. the code is ambigious, can't we find the good translation?

3. The module should load only lot which are not empty

4. The expiration date should be presented as a moment to help user guess easily

5. ~Is the amount necessary?~

6. ~If the quantity is exagerated, the amount content desapper and no way to submit but no message (confusing)~

7. No need of descritption on the receipt paper, because it for intern use the patient don't care about it

8. Instead of gestionnaire stock, let's use just pharmacy because this operation can be performed by an employee who is working at the pharmacy during this specific time. (???)

# TO SERVICE

1. No description is needed for the receipt, it is for an internal use

2. Use pharmacy in signature field (???)

3. the expiration date should be a moment and not a date

# TO DEPOT

1. The two big actions of sending and receiving should be clearly separated and implemented in the system to avoid errors or mistakes. a distribution between depot **X** and **Y** should happen like this : 

* The depot **X** user send items to depot **Y** and update his own depot state (decreasing his stock only)

* The depot **Y** user see the transfer and confirm/ignore/update(?) it, so that, his own depot state is updated by him self (increasing stock only or nothing)

N.B : 
* The sender (depot **X**) can cancel the transfer until the depot **Y** user has accepted it.
* As consequence, no sent items can be used at the destination depot before an explicit confirmation of the depot **Y** user.

2. Changing the unit price and reloading the page after, breaks the module.

# PERTE

1. after using this module, I have found some negative number in the stock

These points can be discussed to find which point to focus on

### PR DESCRIPTION
**Pull request by DedrickEnc**
This PR refactor the stock distribution page and closes #1854

Most of the work is about addressing points listed in #1854 and some refactor in order to reduce the code.
A new component ```bhStockEntryExitType``` has been introduced to remove repeated code found in ```exit.html``` and ```entry.html``` for selecting the type of entry or exit.

The main change is about stock distribution from depot to an other depot (Stock transfer)
Now the stock transfer is a set of two processes which are a stock exit to depot and a stock entry from depot. Both of them are independent in other to avoid mistakes and guaranty the data integrity.
Each time the stock in the depot will match the stock in the system.
 
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through eslint.
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the online review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by DedrickEnc** [comment]
@mbayopanda @sfount 
Can I get a review?

File: client/src/i18n/fr/depot.json [hunk 1]
```diff
@@ -1,8 +1,15 @@
...
-"TITLE":"Gestion des dépôts"}}
\ No newline at end of file
+{
+    "DEPOT": {
+        "ADD_DEPOT": "Ajouter un dépôt",
+        "ALL_DEPOTS": "Tous les dépôts",
+        "DEPOTS_LIST": "Liste des dépôts",
+        "DESCRIPTION": "Ce module vous permet de Lister, Creer, Modifier et Supprimer un dépôt",
+        "EDIT_DEPOT": "Editer les informations d'un dépôt",
+        "ENTITY" : "Depot",
```
Comment (jniles) [inline client/src/i18n/fr/depot.json:8]: `DEPOT.ENTITY` needs to be translated into English.

I haven't checked further, but make sure all your translation keys have a definition in both languages.  If you need any help with the English, feel free to ask.

File: client/src/js/components/bhStockEntryExitType.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        { label: 'purchase', labelKey: 'STOCK.ENTRY_PURCHASE', descriptionKey: 'STOCK_FLUX.FROM_PURCHASE', isEntry: true },
+        { label: 'integration', labelKey: 'STOCK.INTEGRATION', descriptionKey: 'STOCK_FLUX.FROM_INTEGRATION', isEntry: true },
+        { label: 'donation', labelKey: 'STOCK.DONATION', descriptionKey: 'STOCK_FLUX.FROM_DONATION', isEntry: true },
+        { label: 'transfer_reception', labelKey: 'STOCK.RECEPTION_TRANSFER', descriptionKey: 'STOCK_FLUX.FROM_TRANSFER', isEntry: true }
+    ];
+
+    $ctrl.display = function () {
+        if ($ctrl.isEntry) {
+            return $ctrl.reference || '';
+        } else {
```
Comment (jniles) [inline client/src/js/components/bhStockEntryExitType.js]: Since you have `return`, this `else` is unnecessary.  It will increase readability to remove it.

File: client/src/modules/stock/entry/entry.html [hunk 3]
```diff
@@ -21,70 +21,12 @@
...
-                <span ng-show="StockCtrl.movement.entity.type === 'integration'">
-                  {{ StockCtrl.movement.entity.instance.reference }}
-                </span>
-              </div>
-            </div>
-          </div>
-        </button>
-      </div>
-
+      <bh-stock-entry-exit-type
```
Comment (jniles) [inline client/src/modules/stock/entry/entry.html:24]: This was a great idea to make a component :+1:

File: client/src/js/components/bhStockEntryExitType.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+angular.module('bhima.components')
+    .component('bhStockEntryExitType', {
+        templateUrl: 'modules/templates/bhStockEntryExitType.tmpl.html',
+        controller: StockEntryExitTypeController,
+        bindings: {
+            onEntryExitTypeSelectCallback: '&?',
+            reference: '<?',
+            displayName: '<?',
+            isEntry: '<'
```
Comment (jniles) [inline client/src/js/components/bhStockEntryExitType.js]: I think this should probably be `@`.  The `<` is for binding a variable from the parent scope, but it doesn't look like the bindings are ever expected to change.

**Comment by DedrickEnc** [comment]
@IMA-WorldHealth/bhima-core 
Can I get a review for this pull request?

**Comment by DedrickEnc** [comment]
This PR is a refactor based on the functionality review we made, so it should be reviewed.
Can I get a review for this pull request ?
@IMA-WorldHealth/bhima-core

**Comment by jniles** [comment]
@dedrickenc, I can give you a review.  Can you rebase to master first?

**Comment by DedrickEnc** [comment]
Thanks @jniles , doing that

**Comment by DedrickEnc** [comment]
@jniles
The branch is rebased

File: client/src/js/components/bhStockEntryExitType.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
+angular.module('bhima.components')
+    .component('bhStockEntryExitType', {
+        templateUrl: 'modules/templates/bhStockEntryExitType.tmpl.html',
+        controller: StockEntryExitTypeController,
+        bindings: {
+            onEntryExitTypeSelectCallback: '&?',
```
Comment (jniles) [inline client/src/js/components/bhStockEntryExitType.js]: You probably want to make this callback required (`&`).

File: client/src/modules/stock/entry/modals/findPurchase.modal.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ function StockFindPurchaseModalController(Instance, Purchase, Notify,
     { field            : 'reference',
       displayName      : 'TABLE.COLUMNS.REFERENCE',
       headerCellFilter : 'translate',
-      cellTemplate     : 'modules/stock/entry/modals/templates/purchase.reference.tmpl.html' },
+      cellTemplate     : 'modules/stock/entry/modals/templates/purchase_reference.tmpl.html' },
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findPurchase.modal.js:26]: Good catch :+1:

File: client/src/modules/templates/bhStockEntryExitType.tmpl.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    <button id="entry-exit-type-{{type.label}}" class="btn-block panel panel-default segment" type="button" ng-class="{'bg-ima-dark-blue': type.label === $ctrl.selectedEntryExitType.label}"
+        ng-click="$ctrl.selectEntryExitType(type)" name="btn-patient">
+        <div class="panel-body text-center text-ellipsis">
+            <div class="ui lg ima-blue statistic">
+                <div class="value" translate>
+                    {{type.labelKey}}
+                </div>
+                <div>
+                    <div class="uilabel">
+                        <span ng-hide="($ctrl.reference || $ctrl.displayName) && type.label === $ctrl.selectedEntryExitType.label" translate>{{type.descriptionKey}}</span>
```
Comment (jniles) [inline client/src/modules/templates/bhStockEntryExitType.tmpl.html]: This is a pretty complicated expression.  Could you pre-process the data somehow to make this cleaner?  Something like:
```js
$ctrl.entryExistTypeList.forEach(function (btn) {
  btn.hasVisibleDescriptionKey = ($ctrl.reference || $ctrl.displayName) && type.label === $ctrl.selectedEntryExitType.label;
});
```

Also, why do you need to check for `$ctr.reference || $ctrl.displayName`?

File: client/src/modules/templates/bhStockEntryExitType.tmpl.html [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<div class="col-md-3 col-sm-6 col-xs-12" ng-repeat="type in $ctrl.entryExitTypeList | filter:{isEntry : $ctrl.isEntry}">
+    <button id="entry-exit-type-{{type.label}}" class="btn-block panel panel-default segment" type="button" ng-class="{'bg-ima-dark-blue': type.label === $ctrl.selectedEntryExitType.label}"
+        ng-click="$ctrl.selectEntryExitType(type)" name="btn-patient">
```
Comment (jniles) [inline client/src/modules/templates/bhStockEntryExitType.tmpl.html]: What is the `name=btn-patient` for?

File: server/controllers/stock/core.js [hunk 9]
```diff
@@ -61,23 +62,22 @@ function getLots(sqlQuery, parameters, finalClauseParameter) {
...
+          l.expiration_date, BUID(l.inventory_uuid) AS inventory_uuid, i.delay, l.entry_date,
+          i.code, i.text, BUID(m.depot_uuid) AS depot_uuid, d.text AS depot_text, iu.text AS unit_type 
         FROM lot l 
         JOIN inventory i ON i.uuid = l.inventory_uuid 
         JOIN inventory_unit iu ON iu.id = i.unit_id 
         JOIN stock_movement m ON m.lot_uuid = l.uuid AND m.flux_id = ${flux.FROM_PURCHASE} 
         JOIN depot d ON d.uuid = m.depot_uuid 
     `;
-  db.convert(params, ['uuid', 'depot_uuid', 'lot_uuid', 'inventory_uuid']);
+  db.convert(params, ['uuid', 'depot_uuid', 'lot_uuid', 'inventory_uuid', 'document_uuid']);
```
Comment (jniles) [inline server/controllers/stock/core.js:74]: :+1:

File: client/src/modules/templates/bhStockEntryExitType.tmpl.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        <div class="panel-body text-center text-ellipsis">
+            <div class="ui lg ima-blue statistic">
+                <div class="value" translate>
+                    {{type.labelKey}}
+                </div>
+                <div>
+                    <div class="uilabel">
```
Comment (jniles) [inline client/src/modules/templates/bhStockEntryExitType.tmpl.html]: This div should line up with the div above it.

File: server/controllers/stock/core.js [hunk 11]
```diff
@@ -128,6 +130,11 @@ function getLotsDepot(depotUuid, params, finalClause) {
     delete params.status;
   }
 
+  if(params.includeEmptyLot === 'false'){
```
Comment (jniles) [inline server/controllers/stock/core.js]: Is this parameter ever used?  I couldn't find where it was referenced.

If it is used, could we make the parameter selection `0` or `1` instead of `true` or `false`?  That way `Number()` could parse it.

File: server/controllers/stock/core.js [hunk 12]
```diff
@@ -167,16 +174,25 @@ function getLotsDepot(depotUuid, params, finalClause) {
  * @param {object} params - A request query object
  */
 function getLotsMovements(depotUuid, params) {
+  let finalClause;
+
   if (depotUuid) {
     params.depot_uuid = depotUuid;
   }
 
+  if(params.finalClauseParameter){
```
Comment (jniles) [inline server/controllers/stock/core.js]: It looks like you never actually use `finalClauseParameter`.  What was it used for?  Are you sure this doesn't break existing functionality?

File: client/src/modules/stock/entry/modals/findTransfer.modal.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  /** get transfer document */
+  function showReceipt(uuid) {
+    Receipts.stockExitDepotReceipt(uuid, true);
+  }
+
+  /* ======================= End Grid ======================================== */
+  StockService.movements.read(null, {
+    entity_uuid: data.depot_uuid,
+    is_exit: 1,
+    finalClauseParameter: 'GROUP BY document_uuid',
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.js]: This seems like it might be vulnerable to SQLi.  Even if it isn't, we shouldn't be writing SQL on the client.

File: client/src/modules/stock/entry/modals/templates/transfer_view.tmpl.html [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<div class="ui-grid-cell-contents">
+  <a ng-click="grid.appScope.showReceipt(row.entity.document_uuid)" href translate>TABLE.COLUMNS.RECEIPT</a>
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/templates/transfer_view.tmpl.html:2]: This link is broken on the test data ... I'm not sure if the feature is broken or if it's just a fluke.  Could you investigate?

File: client/src/modules/stock/exit/exit.html [hunk 15]
```diff
@@ -18,196 +21,117 @@
...
-                <span ng-show="StockCtrl.movement.entity.type === 'patient'">
-                  {{ StockCtrl.movement.entity.instance.reference }} - {{ StockCtrl.movement.entity.instance.display_name }}
-                </span>
-              </div>
-            </div>
-          </div>
-        </button>
+      <!-- destination -->
+      <div class="row">
+        <bh-stock-entry-exit-type
```
Comment (jniles) [inline client/src/modules/stock/exit/exit.html:26]: :+1:

File: client/src/modules/stock/entry/modals/findTransfer.modal.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+  function rowSelectionCallback(row) {
+    vm.selectedRow = row.entity;
+  }
+
+  /** toggle filter */
+  function toggleFilter() {
+    vm.filterEnabled = !vm.filterEnabled;
+    vm.gridOptions.enableFiltering = vm.filterEnabled;
+    vm.gridApi.core.notifyDataChange(uiGridConstants.dataChange.COLUMN);
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.js]: I think this should be `notifyDataChange(uiGridConstants.dataChange.OPTION);`.  [Reference documentation](http://ui-grid.info/docs/#/api/ui.grid.service:uiGridConstants#properties_datachange).

File: client/src/modules/stock/entry/modals/findTransfer.modal.js [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    {
+      field: 'action',
+      displayName: '',
+      enableFiltering: false,
+      enableSorting: false,
+      cellTemplate: 'modules/stock/entry/modals/templates/transfer_view.tmpl.html'
+    }
+  ];
+
+  vm.gridOptions.columnDefs = columns;
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.js:44]: This grid should turn off the column menu as well.

File: test/end-to-end/shared/components/bhFiscalPeriodSelect.js [hunk 18]
```diff
@@ -21,8 +21,5 @@ module.exports = {
 
     opts = bhFiscalPeriod.element(by.model('$ctrl.selectedPeriodTo'));
     opts.element(getValue(periodTo_id)).click();
-
-    // click away to free the space
```
Comment (jniles) [inline test/end-to-end/shared/components/bhFiscalPeriodSelect.js:25]: Out of curiosity, why was this removed?

File: test/end-to-end/stock/stock.exit.page.js [hunk 19]
```diff
@@ -56,7 +56,7 @@ function StockExitPage() {
    * @method setLoss
    */
   page.setLoss = function setLoss() {
-    element(by.css('[name="btn-loss"]')).click();
+    components.stockEntryExitType.set('entry-exit-type-loss');
```
Comment (jniles) [inline test/end-to-end/stock/stock.exit.page.js]: It seems like writing `entry-exit-type` over and over in each ID is a duplication of effort.  Can't it just be `.set('loss')` or something similar?

File: client/src/modules/stock/entry/modals/findTransfer.modal.js [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+  vm.gridOptions.multiSelect = false;
+  vm.gridOptions.enableFiltering = vm.filterEnabled;
+  vm.gridOptions.onRegisterApi = onRegisterApi;
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.js:47]: It would be good to use `flatEntityAccess` and `fastWatch` if possible.

File: client/src/modules/stock/exit/templates/expiration.tmpl.html [hunk 21]
```diff
@@ -1,3 +1,4 @@
 <div ng-if="row.entity.lot.uuid" class="ui-grid-cell-contents">
-  {{ row.entity.lot.expiration_date | date }}
+  {{grid.appScope.util.getTimeFromToday(row.entity.lot.expiration_date, 'days')}}
```
Comment (jniles) [inline client/src/modules/stock/exit/templates/expiration.tmpl.html]: I think `angular-moment` does this automatically.  I think the [am-time-ago](https://github.com/urish/angular-moment#am-time-ago-directive) directive is what you need.

File: server/controllers/stock/index.js [hunk 22]
```diff
@@ -176,58 +176,43 @@ function normalMovement(document, params) {
...
-      depot_uuid    : db.bid(parameters.to_depot),
+      depot_uuid    : isExit ? db.bid(parameters.from_depot) : db.bid(parameters.to_depot),
       document_uuid : db.bid(document.uuid),
       quantity      : lot.quantity,
       unit_cost     : lot.unit_cost,
       date          : document.date,
-      entity_uuid   : null,
-      is_exit       : 0,
-      flux_id       : core.flux.FROM_OTHER_DEPOT,
+      entity_uuid   : isExit ? db.bid(parameters.to_depot) : db.bid(parameters.from_depot),
```
Comment (jniles) [inline server/controllers/stock/index.js]: Instead of doing this on every turn of the loop, can you do something like this:
```js
const depot_uuid =  db.bid(isExit ? parameters.from_depot : parameters.to_depot);
const entity_uuid = db.bid(isExit ? parameters.to_depot : parameters.from_depot);
const is_exit = isExit ? 1 : 0;
const flux_id = isExit ? core.flux.TO_OTHER_DEPOT : core.flux.FROM_OTHER_DEPOT;

parameters.lots.forEach((lot) => {
    record = {
      depot_uuid,
      is_exit,
      entity_uuid,
      // ... etc ...
    };
});
```

File: server/controllers/stock/reports/stock_exit_depot.receipt.handlebars [hunk 23]
```diff
@@ -36,15 +36,15 @@
       </div>
       <div class="col-xs-6">
         <h4>{{translate 'STOCK.TO'}}</h4>
-        <span class="text-capitalize">{{translate 'STOCK.DEPOT'}}</span>: <strong>{{entry.details.depot_name}}</strong> <br>
-        <span class="text-capitalize">{{translate 'FORM.LABELS.DOCUMENT'}}</span>: <strong>{{entry.details.document_reference}}</strong> <br>
+        <span class="text-capitalize">{{translate 'STOCK.DEPOT'}}</span>: <strong>{{exit.details.otherDepotName}}</strong> <br>
```
Comment (jniles) [inline server/controllers/stock/reports/stock_exit_depot.receipt.handlebars:39]: :dancer:

File: test/end-to-end/stock/stock.adjustment.spec.js [hunk 24]
```diff
@@ -52,4 +52,4 @@ function StockAdjustmentTests() {
   });
 }
 
-describe('Stock Adustment Test', StockAdjustmentTests);
+describe('Stock Adjustment Test', StockAdjustmentTests);
```
Comment (jniles) [inline test/end-to-end/stock/stock.adjustment.spec.js:55]: :+1:

File: client/src/modules/stock/entry/modals/findTransfer.modal.html [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+      </div>
+    </div>
+  </div>
+
+  <div class="modal-footer">
+    <button type="button" class="btn btn-default" ng-click="$ctrl.cancel()" data-method="cancel" translate>
+      FORM.BUTTONS.CLOSE
+    </button>
+
+    <bh-loading-button loading-state="FindForm.$loading">
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.html]: I can submit this without any choices and it loads every lot.  Is this intended behavior?

File: client/src/modules/stock/entry/modals/findTransfer.modal.html [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+    <div class="row">
+      <div class="col-xs-12">
+        <div
+          id="TransferGrid"
+          ui-grid="$ctrl.gridOptions"
+          ui-grid-auto-resize
+          ui-grid-resize-columns
+          ui-grid-selection>
+          <bh-grid-loading-indicator
+            loading-state="$ctrl.loading"
```
Comment (jniles) [inline client/src/modules/stock/entry/modals/findTransfer.modal.html:35]: Is this loading property ever defined?  I couldn't find it.

**Review by jniles — CHANGES_REQUESTED** [review]
@DedrickEnc, I've done an initial review.  There's a few broken pieces and a few optimizations to make.

Let me know if you have any questions.

File: client/src/modules/templates/bhStockEntryExitType.tmpl.html (hunk 7 above)
Comment (DedrickEnc) [reply client/src/modules/templates/bhStockEntryExitType.tmpl.html]: We skip it for now

File: server/controllers/stock/core.js (hunk 12 above)
Comment (DedrickEnc) [reply server/controllers/stock/core.js]: it is used to init ```finalClause``` variable at the next line

File: test/end-to-end/shared/components/bhFiscalPeriodSelect.js (hunk 18 above)
Comment (DedrickEnc) [reply test/end-to-end/shared/components/bhFiscalPeriodSelect.js:25]: Did it since a long time but If my memory serves me well, the component was inside a modal so no way to click on the logo which is behind the modal.

Comment (jniles) [reply test/end-to-end/shared/components/bhFiscalPeriodSelect.js:25]: Aha.  That makes sense. :+1:

Comment (sfount) [reply test/end-to-end/shared/components/bhFiscalPeriodSelect.js:25]: l'anglais 👌

**Comment by DedrickEnc** [comment]
@jniles
Changes has been integrated

File: client/src/js/components/bhStockEntryExitType.js [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+
+StockEntryExitTypeController.$inject = [];
+
+/**
+ * Stock Entry Exit Type component
+ */
+function StockEntryExitTypeController() {
+    var $ctrl = this;
+    $ctrl.selectedEntryExitType = null;
+    $ctrl.isEntry === 'true';
```
Comment (jniles) [inline client/src/js/components/bhStockEntryExitType.js]: Does this actually work?

I think you might be missing an assignment - this is just an evaluation.  Maybe...
```js
$ctrl.isEntry = $ctrl.isEntry === 'true'
```
?

Either way, this should probably go in an `$onInit` method to make sure the bindings are correctly initialized.

**Review by jniles — COMMENTED** [review]
Just one more thing.  After that, LGTM.

**Review by jniles — APPROVED** [review]
